### PR TITLE
Add gcs object manager

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
Merge object_locator and object_info_handler_impl into gcs_object_manager.
gcs_object_manager caches data in memory instead of writing redis.
If gcs_server is invalid,  we need to recover the object location data from the raylet, which will be implemented in the next pr.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Review by raulchen — COMMENTED** [review]
Can you add a unit test for `GcsObjectManager`? Just keep the logic of `object_locator_test`.

**Comment by ffbin** [comment]
I have add a unit test for GcsObjectManager, thanks.

File: src/ray/gcs/gcs_server/gcs_object_manager.h [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  void HandleRemoveObjectLocation(const rpc::RemoveObjectLocationRequest &request,
+                                  rpc::RemoveObjectLocationReply *reply,
+                                  rpc::SendReplyCallback send_reply_callback) override;
+
+ protected:
+  /// Add a location of objects.
+  ///
+  /// \param node_id The object location that will be added.
+  /// \param object_ids The ids of objects which location will be added.
+  void AddObjectsLocation(const ClientID &node_id,
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_object_manager.h:62]: this function is only a helper function for test? If so, why not just call `AddObjectLocation` multiple times? perf should be fine for test.

File: src/ray/gcs/gcs_client/global_state_accessor.h [hunk 2]
```diff
@@ -51,6 +51,21 @@ class GlobalStateAccessor {
   /// protobuf function.
   std::vector<std::string> GetAllJobInfo();
 
+  /// Get information of all objects from GCS Service.
+  ///
+  /// \return All object info. To support multi-language, we serialized each
+  /// ObjectTableData and returned the serialized string. Where used, it needs to be
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/global_state_accessor.h]: fixed

File: src/ray/gcs/gcs_client/global_state_accessor.h [hunk 3]
```diff
@@ -51,6 +51,21 @@ class GlobalStateAccessor {
...
+  /// deserialized with protobuf function.
+  std::vector<std::string> GetAllObjectInfo();
+
+  /// Get information of object from GCS Service.
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/global_state_accessor.h]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.h [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  ///
+  /// \param object_id The id of object to lookup.
+  /// \return Object locations.
+  absl::flat_hash_set<ClientID> GetObjectLocations(const ObjectID &object_id)
+      LOCKS_EXCLUDED(mutex_);
+
+  /// Remove a node.
+  ///
+  /// \param node_id The node that will be removed.
+  void RemoveNode(const ClientID &node_id) LOCKS_EXCLUDED(mutex_);
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.h]: fixed

File: python/ray/includes/global_state_accessor.pxi [hunk 5]
```diff
@@ -22,3 +26,12 @@ cdef class GlobalStateAccessor:
 
     def get_job_table(self):
         return self.inner.get().GetAllJobInfo()
+
+    def get_object_table(self):
+        return self.inner.get().GetAllObjectInfo()
+
+    def get_object_info(self, object_id):
+        object_info = self.inner.get().GetObjectInfo(CObjectID.FromBinary(object_id.binary()))
+        if object_info.strip() == '':
```
[the remark this answers is not in the document]
Comment (ffbin) [reply python/ray/includes/global_state_accessor.pxi]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.h [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  /// \param create_if_not_exist Whether to create a new one if not exist.
+  /// \return ObjectSet *
+  GcsObjectManager::ObjectSet *GetNodeHoldObjectSet(const ClientID &node_id,
+                                                    bool create_if_not_exist = false)
+      EXCLUSIVE_LOCKS_REQUIRED(mutex_);
+
+  mutable absl::Mutex mutex_;
+
+  /// Mapping from object id to object locations.
+  absl::flat_hash_map<ObjectID, LocationSet> object_to_locations_ GUARDED_BY(mutex_);
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.h]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.h [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  void AddObjectsLocation(const ClientID &node_id,
+                          const absl::flat_hash_set<ObjectID> &object_ids)
+      LOCKS_EXCLUDED(mutex_);
+
+  /// Add a location of an object.
+  ///
+  /// \param object_id The id of object which location will be added.
+  /// \param node_id The object location that will be added.
+  void AddObjectLocation(const ObjectID &object_id, const ClientID &node_id)
+      LOCKS_EXCLUDED(mutex_);
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.h]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.h [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+  /// Get object's locations.
+  ///
+  /// \param object_id The id of object to lookup.
+  /// \return Object locations.
+  absl::flat_hash_set<ClientID> GetObjectLocations(const ObjectID &object_id)
+      LOCKS_EXCLUDED(mutex_);
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.h]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.cc [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+          gcs::CreateObjectLocationChange(node_id, true)->SerializeAsString(), nullptr));
+      RAY_LOG(DEBUG) << "Finished adding object location, job id = "
+                     << object_id.TaskId().JobId() << ", object id = " << object_id
+                     << ", node id = " << node_id << ", task id = " << object_id.TaskId();
+    } else {
+      RAY_LOG(ERROR) << "Failed to add object location: " << status.ToString()
+                     << ", job id = " << object_id.TaskId().JobId()
+                     << ", object id = " << object_id << ", node id = " << node_id;
+    }
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.cc:84]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.h [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  /// Remove a node.
+  ///
+  /// \param node_id The node that will be removed.
+  void RemoveNode(const ClientID &node_id) LOCKS_EXCLUDED(mutex_);
+
+  /// Remove object's location.
+  ///
+  /// \param object_id The id of the object which location will be removed.
+  /// \param node_id The location that will be removed.
+  void RemoveObjectLocation(const ObjectID &object_id, const ClientID &node_id)
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.h]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.cc [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+  if (object_locations) {
+    return *object_locations;
+  }
+  return absl::flat_hash_set<ClientID>{};
+}
+
+void GcsObjectManager::RemoveNode(const ClientID &node_id) {
+  absl::MutexLock lock(&mutex_);
+
+  ObjectSet node_hold_objects;
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.cc]: fixed

File: src/ray/gcs/gcs_server/gcs_object_manager.h (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_object_manager.h:65]: If the GCS server restarts, this function is used to reload data from storage.

**Review by raulchen — APPROVED** [review]

**Pull request merged**